nimbus/src/lib.rs

```diff
@@ -162,9 +194,20 @@ pub struct EnrolledExperiment {
     pub branch_slug: String,
 }
 
+// XXX: In the future it would be nice if this lived in its own versioned crate so that
+// the schema could be decoupled from the sdk so that it can be iterated on while the
+// sdk depends on a particular version of the schema through the cargo.toml.
+
+#[derive(Deserialize, Serialize, Debug, Clone, PartialEq)]
+#[serde(untagged)]
```

Comment:
I'm concerned about the use of `untagged` here making the system fragile and harder to reason about.

For example, suppose that we define a "2.0.0" schema version. Naively, I'd expect us to want to add a `V2(V2_0_0)` variant to this enum. But what if the JSON data in `V2_0_0` also happens to be parseable as a valid `V1_0_0` struct (perhaps because it adds new fields that the `V1_0_0` parser will ignore)? Whether we end up with a `V1` or a `V2` will depend on which order the variants are declared in the enum, which is non-obvious.

Alternately, we could evolve the `V1_0_0` struct in such a way that it can represent both "1.0.0" and "2.0.0" experiment data, and inspect the `schema_version` attribute at runtime to decide which semantics to impose. I think this would make the consuming code harder to reason about because you might naturally read `Experiment::V1` and assume that it's a "1.x.y" experiment.

It might be cleaner to instead bypass serde's magic here and implement our own switching on the `schema_version` field, something like:

```
pub fn parse_experiment_json(data: &serde_json::Value) -> Result<Experiment> {
  let schema_version = data.get("schema_version").ok_or(Error:InvalidExperimentResponse)?;
  if V1_SEMVER_MATCH(schema_version) {
    return Experiment::V1(serde_json::from_value::<V1_0_0>(data)?)
  }
  if V2_SEMVER_MATCH(schema_version) {
    return Experiment::V2(serde_json::from_value::<V2_0_0>(data)?)
  }
  return Err(Error:UnsupportedSchemaVersion)
}
```

It's a little more boilerplate but I think it would make it more obvious that we're getting the intended result.